### ISSUE TITLE
Seperately trained on FashionIQ subsets?

### ISSUE THREAD

**Issue by BrandonHanx** [postBG/CosMo.pytorch#3]
Hi, @numpee . Another question please.

I wonder did you train three models on FashionIQ dress/toptee/shirt separately?

In other words, are the results shown in Table 1 from one model or three models?

**Comment by numpee** [comment]
Yes, those are three separate datasets and we train a different model on each.
You can change the dataset by using the `--dataset={fashionIQ_dress|fashionIQ_shirt|fashionIQ_toptee}` flag

**Comment by BrandonHanx** [comment]
Okay, thanks for your explanation.

I am not familiar with the experimental settings of FashionIQ...  
Did VAL and TIRG train their models separately? 
I guess they trained on all subsets together while tested the model on different subsets separately...

BTW, one [AAAI2021 work](https://ojs.aaai.org/index.php/AAAI/article/view/16271), also from SNU,  trained only one model on all subsets.

**Comment by numpee** [comment]
@BrandonHanx 
Hmm this is interesting. We trained separately on each dataset, but as you mentioned, it seems other works may have trained a single model on all datasets and evaluated them separately.
We just tried training a single model on all 3 datasets, and found that the average R@10 and R@50 for this setting is actually better than the one reported in our paper. It seems like the performance becomes better since there is more training data. 
We're still running some experiments, so we'll give you more updated results when the experiments are done. Thanks for pointing this out!

**Comment by BrandonHanx** [comment]
Thanks! Looking forward to your higher performance.

**Comment by BrandonHanx** [comment]
Hi, @numpee .

I tried to reproduce your results in the paper (with original split). 
I used the vocab you shared and the default configs, but only got 19.81 R@10 on FashionIQ dress, while yours reported in supp is 21.39.

Any other settings need to be addressed? Please give me some hints.
This is my wandb logs: https://wandb.ai/brandon-han/CoSMo.pytorch/runs/rmh3o7vm/overview?workspace=user-brandon-han

**Comment by numpee** [comment]
Hi @BrandonHanx,

Here are the results using all 3 subsets to train a single model. BTW, this is using the VAL evaluation method:

|                                 | Dress R@10 | Dress R@50 | Toptee R@10 | Toptee R@50 | Shirt R@10 | Shirt R@50 | Avg R@10 | Avg R@50 |
|---------------------------------|:----------:|:----------:|:-----------:|:-----------:|:----------:|:----------:|:--------:|:--------:|
| CoSMo (Paper)                   |    25.64   |    50.30   |    29.21    |    57.46    |    24.90   |    49.18   |   26.58  |   52.31  |
| CoSMo  (One model for all exps) |    26.45   |    52.43   |    31.95    |    62.09    |    26.94   |    52.99   |   28.45  |   55.84  |

Interestingly, training on all subsets leads to better performance than what we reported in the paper.
If you want to try it out yourself, one easy method is to modify the `train_dataset_factory` function in `data/__init__.py` to load all three datasets, then use `torch.utils.data.ConcatDataset` to combine all three datasets. One caveat is that you would need to train this model 3 times, each with a different evaluation dataset setting. BTW, for this setting, you may want to reduce the `decay_step` to 20 and the number of total epochs as well. We will add code for this unified setting in the near future (it's a bit messy right now 😢) since we have a few other things going on right now.   

Regarding reproducibility, you may want to try it on multiple random seeds, or also try it with the VAL evaluation method (check the README on how to enable this). You can set a randomly-generated random seed by setting the `--random_seed=-1`.

**Comment by BrandonHanx** [comment]
Hi @numpee ,

Thanks for your detailed results and useful hints.  This issue is solved for me.

**Comment by BrandonHanx** [comment]
Hi @numpee ,

I still cannot reproduce similar results with those reported in your paper.

Are you using the config files provided in this repo? If not, could you please share your config?

**Comment by numpee** [comment]
Hm, we checked reproducibility before we uploaded the code. Have you tried with the other datasets as well, other than dress? Anyways, we'll update the code to support multi-dataset setting soon.
